keanu-project/src/main/java/io/improbable/keanu/plating/loop/LoopBuilder.java

```diff
@@ -0,0 +1,180 @@
+     * If the maximum loop count is exceeded, it will log a warning instead of throwing
+     *
+     * @return self
+     */
+    public LoopBuilder dontThrowWhenMaxCountIsReached() {
+        this.throwWhenMaxCountIsReached = false;
+        return this;
+    }
+
+    /**
+     * An optional method to add custom mappings
+     *
+     * @param inputLabel  - the label assigned to the ProxyVertex in frame t
+     * @param outputLabel - the label assigned to a Vertex in frame t-1 which will become the ProxyVertex's parent
+     * @return self
+     */
+    public LoopBuilder mapping(VertexLabel inputLabel, VertexLabel outputLabel) {
+        customMappings.put(inputLabel, outputLabel);
+        return this;
+    }
+
+    /**
+     * A mandatory method to specify the condition
+     *
+     * @param conditionSupplier - a lambda that creates and returns a new BoolVertex
+     * @return the next stage builder
+     */
+    public LoopBuilder2 whilst(Supplier<BoolVertex> conditionSupplier){
+        return whilst(plate -> conditionSupplier.get());
+    }
+
+    /**
+     * A mandatory method to specify the condition
+     *
+     * @param conditionFunction - a lambda that takes the current Plate and creates and returns a new BoolVertex
+     * @return the next stage builder
+     */
+    public LoopBuilder2 whilst(Function<Plate, BoolVertex> conditionFunction){
+        return new LoopBuilder2(initialState, conditionFunction, customMappings.build(), maxLoopCount, throwWhenMaxCountIsReached);
+    }
+
+    public class LoopBuilder2 {
+        private final Collection<Vertex> initialState;
+        private final Function<Plate, BoolVertex> conditionFunction;
+        private final Map<VertexLabel, VertexLabel> customMappings;
+        private final int maxLoopCount;
+        private final boolean throwWhenMaxCountIsReached;
+        private final VertexLabel VALUE_OUT_WHEN_ALWAYS_TRUE_LABEL = new VertexLabel("loop_value_out_when_always_true");
+        private final VertexLabel VALUE_IN_WHEN_ALWAYS_TRUE_LABEL = PlateBuilder.proxyFor(VALUE_OUT_WHEN_ALWAYS_TRUE_LABEL);
+        private final VertexLabel LOOP_LABEL = new VertexLabel("loop");
+
+
+        LoopBuilder2(Collection<Vertex> initialState, Function<Plate, BoolVertex> conditionFunction, Map<VertexLabel, VertexLabel> customMappings, int maxLoopCount, boolean throwWhenMaxCountIsReached){
+            this.initialState = setInitialState(initialState);
+            this.conditionFunction = conditionFunction;
+            this.customMappings = customMappings;
+            this.maxLoopCount = maxLoopCount;
+            this.throwWhenMaxCountIsReached = throwWhenMaxCountIsReached;
+        }
+
+        private ImmutableList<Vertex> setInitialState(Collection<Vertex> initialState){
+            Vertex valueOutWhenAlwaysTrue = null;
+
+            try {
+                List<Vertex> outputVertices = initialState.stream().filter(v -> Loop.VALUE_OUT_LABEL.equals(v.getLabel())).collect(Collectors.toList());
```

Comment:
done - nice!

Reply:
`findFirst` is neat, however `getOnlyElement` is a stricter check as it will throw if there is more than one. But anyway, I can address your point by inlining `outputVertices` so it's a one-liner:

    Vertex<?> outputVertex = Iterables.getOnlyElement(initialState.stream().filter(v -> Loop.VALUE_OUT_LABEL.equals(v.getLabel())).collect(Collectors.toList()));

Reply:
Nice